### ISSUE TITLE
Reproduce Cora embedding from VGAE paper by Kipf & Welling 2016

### ISSUE THREAD

**Issue by wehlutyk** [ixxi-dante/nw2vec#44]
Four things, each to be visualised using downscaling and projections (the paper doesn't specify the method used to produce their 2D figure from 16D embeddings):
- [x] run their `gae` implementation directly to reproduce their own procedure
- [x] run our implementation with their parameters:
  - no `u` parameter for embedding covariance matrix (remove the layer, and set `u` to 0 in the codec)
  - simple scalar product decoder
  - intermediate layer 32 dims, embedding layer 16 dims
  - train for 200 iterations
  - get the loading code from their `gae` implementation
- [x] reactivate kernel in scalar product
- [x] add the 2-layer decoder we've been using
- [ ] add the `u` term in the embedding covariance matrix

Note that the current implementation mimics their `gae` closer than what we'll have after implementing #7, so do this before.

**Comment by wehlutyk** [comment]
The vanilla `gae` works fine and reproduces the figure in the paper. The paper doesn't mention the downscaling method they use, so I do a generic sklearn multidimensional scaling. I suspect using t-SNE could improve the visualisations (though the original paper doesn't seem to use that, as a quick test created a very different figure), but I didn't try it.

Our implementation gives pretty much the same thing, with 2 caveats:
- it's about 10 times slower: on my laptop I get ~3.3s per epoch, vs. .3-.5s for the vanilla `gae`
- it doesn't reach the same final performance (though this depends on the architecture): in 200 epochs, the `gae` reaches a final loss of .43, and we reach .6

Reproduction of original `gae` figure:

![Reproduction of original `gae` figure](https://user-images.githubusercontent.com/1823010/45885285-530f2880-bdb6-11e8-9b21-d5fbbe064554.png)

`nw2vec` (our implementation) and the same architecture:

![`nw2vec` (our implementation) and the same architecture](https://user-images.githubusercontent.com/1823010/45885323-6e7a3380-bdb6-11e8-89b2-2e43d8138c6e.png)

`nw2vec` and the same architecture + a kernel in the scalar product decoding:

![`nw2vec` and the same architecture + a kernel in the scalar product decoding](https://user-images.githubusercontent.com/1823010/45885813-bb123e80-bdb7-11e8-9a6e-7b16a6a18813.png)

`nw2vec` and the same architecture + an intermediate layer in decoding (from here on it's normal that this is less separated, since the intermediate layer allows for more computation before the scalar product):

![`nw2vec` and the same architecture + an intermediate layer in decoding](https://user-images.githubusercontent.com/1823010/45888201-c700ff00-bdbd-11e8-86c1-5964e8845248.png)

`nw2vec` and the same architecture + a kernel in the scalar product decoding + an intermediate layer in decoding:

![`nw2vec` and the same architecture + a kernel in the scalar product decoding + an intermediate layer in decoding](https://user-images.githubusercontent.com/1823010/45886938-b13e0a80-bdba-11e8-897c-0b7dbe6a3d37.png)

`nw2vec` and the same architecture + a kernel in the scalar product decoding + an intermediate layer in decoding + non-orthogonal gaussian embedding (`u`):

![`nw2vec` and the same architecture + a kernel in the scalar product decoding + an intermediate layer in decoding + non-orthogonal gaussian embedding (`u`)](https://user-images.githubusercontent.com/1823010/45889467-d3d32200-bdc0-11e8-842c-e6461be0e48d.png)

The embeddings seem to become much more mixed when putting more logic in the decoder (makes sense, since we need that logic to interpret the embeddings), which questions whether we should put any logic at all in the decoder (i.e., maybe only if it's useful to be able to combine the network and feature dimensions).

---

Conclusion:

- our implementation works
- logic in the decoder may not be useful (simple is better, I like this!)
- we need to optimise more

Next steps:
- the small tests of #14
- performance optimisations! #23 #24 #25 (and compare to vanilla `gae` profiles)

**Comment by wehlutyk** [comment]
About the difference in final training values:
- [x] `nw2vec` uses neighbour sampling → removed for comparison
- [x] using `float64` instead of `float32` makes for better comparisons
- [x] `nw2vec` has a regularizer on the inner product decoder kernel → removed for comparison
- [x] `nw2vec` normalised adj loss by the density of the adj *with* diagonal elements, when `gae` uses the density of the adj *without* diagonal elements. It seems that what we want is indeed the density *without* the diagonal elements (which corresponds to the real links). → fixed

`gae` encodes `σ` on embeddings when `nw2vec` encodes the inverse of the covariance matrix, i.e. `1 / σ²` → making `gae` encode `1 / σ²` reduces the difference in final loss, but this was only useful when looking for a more general parametrisation of the covariance matrix (following Rezende et al. 2014). It seems better to directly encode `σ`, and recheck that performance is comparable with `gae` → #53.

About the difference in speed: it seems that because of a more general approach (in particular for minibatch), we repeat many operations for each epoch, making the whole thing much less optimised. I'll explore that in depth in the Performance project.
